service/gov_pay.go

```diff
@@ -135,6 +145,86 @@ func (gp *GovPayService) GetGovPayPaymentDetails(paymentResource *models.Payment
 	return paymentDetails, Success, nil
 }
 
+// GetGovPayRefundSummary gets refund summary of a GovPay payment
+func (gp *GovPayService) GetGovPayRefundSummary(req *http.Request, id string) (*models.PaymentResourceRest, *models.RefundSummary, ResponseType, error) {
+	// Get PaymentSession for the GovPay call
+	paymentSession, response, err := gp.PaymentService.GetPaymentSession(req, id)
+	if err != nil {
+		err = fmt.Errorf("error getting payment resource: [%v]", err)
+		log.ErrorR(req, err)
+		return nil, nil, response, err
+	}
+
+	govPayResponse, err := callGovPay(gp, paymentSession)
+	if err != nil {
+		err = fmt.Errorf("error getting payment information from gov pay: [%v]", err)
+		log.ErrorR(req, err)
+
+		return nil, nil, Error, err
+	}
+
+	switch govPayResponse.RefundSummary.Status {
+	case RefundUnavailable:
+		err = errors.New("cannot refund the payment - check if the payment failed")
+		return nil, nil, InvalidData, err
+	case RefundFull:
+		err = errors.New("cannot refund the payment - the full amount has already been refunded")
+		return nil, nil, InvalidData, err
+	case RefundPending:
+		err = errors.New("cannot refund the payment - the user has not completed the payment")
+		return nil, nil, InvalidData, err
+	}
+
+	// Return the payment session info and refund summary
+	return paymentSession, &govPayResponse.RefundSummary, Success, nil
+}
+
+// CreateRefund creates a refund in GovPay
+func (gp *GovPayService) CreateRefund(paymentResource *models.PaymentResourceRest, refundRequest *models.CreateRefundGovPayRequest) (*models.CreateRefundGovPayResponse, ResponseType, error) {
+	requestBody, err := json.Marshal(refundRequest)
+	if err != nil {
+		return nil, Error, fmt.Errorf("error reading refund GovPayRequest: [%s]", err)
+	}
+
+	request, err := http.NewRequest("POST", paymentResource.MetaData.ExternalPaymentStatusURI+"/refunds", bytes.NewBuffer(requestBody))
+	if err != nil {
+		return nil, Error, fmt.Errorf("error generating request for GovPay: [%s]", err)
+	}
+
+	if paymentResource.Costs[0].ClassOfPayment[0] == "penalty" {
+		request.Header.Add("authorization", "Bearer "+gp.PaymentService.Config.GovPayBearerTokenTreasury)
+	}
+	if paymentResource.Costs[0].ClassOfPayment[0] == "data-maintenance" ||
```

Comment:
else if?

Reply:
Handled identically to GenerateNextURLGovPay method in gov_pay.go. @divya051988 , @dalebradley can you confirm if this is the correct way to handle it?

Reply:
Yeah makes sense to add else if and log error message or something - `class of payment not set` ?